Custom Form Input Cleanup

Allcharles opened the pull request:
# Custom Form Inputs

Cleanup form behaviour by creating a few custom formly inputs which handle specific form validations.

## Changes

- Modified timezone input
  - Added some visual feedback to help users identify when a timezone is properly set
- Create password input
  - This creates a password, and password confirmation field which validate each other
- Created location input
  - This creates a latitude and longitude field
  - Both fields are validated to ensure they are either both set, or unset. And that the values are valid (-90 -> 90 and -180 -> 180)
  - Displays position of lat/lng on map so users can verify it is in the right location
    - Ended up not making it so user can drag marker, may be revisited later if requested as a feature
- Modified `RenderFieldComponent` to better handle images
- Extracted custom formly inputs into a singular module
- Extracted `IndicatorComponent` into a module
- Added `XOR` type
- Fixed issue in baw interceptor where incoming error messages were not being converted to camelCase

## Problems

- Unit tests are broken (separate PR)

## Issues

Closes #341 
Related to #51 

## Visual Changes

### Lat/Lng Input

![image](https://user-images.githubusercontent.com/3955116/88493508-db833380-cff4-11ea-941b-ca3f47fd7f00.png)
![image](https://user-images.githubusercontent.com/3955116/88493466-a971d180-cff4-11ea-9c9a-37c97ee6c0fb.png)
![image](https://user-images.githubusercontent.com/3955116/88493532-f786d500-cff4-11ea-8a7d-46d3bec135ef.png)

### Password Input

![image](https://user-images.githubusercontent.com/3955116/88493578-3452cc00-cff5-11ea-9e15-05c836fe255d.png)
![image](https://user-images.githubusercontent.com/3955116/88493581-39b01680-cff5-11ea-9822-08be75f011d5.png)
![image](https://user-images.githubusercontent.com/3955116/88493587-40d72480-cff5-11ea-97fa-24cf9348e54d.png)
![image](https://user-images.githubusercontent.com/3955116/88493592-46cd0580-cff5-11ea-899b-cb273e65e0ed.png)


### Timezone Input

![image](https://user-images.githubusercontent.com/3955116/88493558-17b69400-cff5-11ea-908b-ec1e2c1d32da.png)
![image](https://user-images.githubusercontent.com/3955116/88493554-138a7680-cff5-11ea-8a7a-62dfa98026f4.png)
![image](https://user-images.githubusercontent.com/3955116/88493566-1f763880-cff5-11ea-898a-12c978bcaa75.png)
![image](https://user-images.githubusercontent.com/3955116/88493571-243aec80-cff5-11ea-86e6-661b70a32097.png)

### `RenderFieldComponent`

#### Handling array of images

![image](https://user-images.githubusercontent.com/3955116/88493640-94497280-cff5-11ea-9a61-ed8ee2bc0243.png)
![image](https://user-images.githubusercontent.com/3955116/88493661-a62b1580-cff5-11ea-9be2-1979262705f9.png)

#### Handling single image

![image](https://user-images.githubusercontent.com/3955116/88493671-b8a54f00-cff5-11ea-83e3-5eada66bfb17.png)

Allcharles committed:
Working site location and timezone fields

Update breaks registration form, and removes lat/lng checking in site forms. To fix this, I will need to make custom inputs for location and password verification

Allcharles committed:
Removed fieldGroups

- Removed all fieldGroups
- Created a couple files to handle location ahd password verifications
- Make custom-inputs module to group custom formly inputs

Allcharles committed:
Working location input

Allcharles committed:
Cleanup how render-field handles images

Allcharles committed:
Created password confirmation input

Allcharles committed:
Added map to location-input

Allcharles committed:
Change Detection Loop Bug

Fixed a bug where infinite change detection would occur on the `MapComponent`. This stemmed from the `Site.getMapMarker()` function which returns a new object each time it is called. If the function is called directly into a components HTML, change detection will continue to evaluate it as a new object and infinitely load eventually lagging out the page. To solve, I've created a `sanitizeMapMarkers()` function which can be used to fix change detection for single, and multiple marker use cases.

Allcharles committed:
Added visual feedback to timezone input

Added country name to give user visual feedback when a timezone has been selected for a form. Also extracted the IndicatorComponent so that it can be used as a module.

Allcharles committed:
Remove unneeded locationValidator

Allcharles committed:
Final cleanup

Allcharles committed:
Merge branch 'iana-timezone' into remove-form-validators

atruskie commented on src/app/models/AssociationDecorators.ts:
```diff
@@ -7,27 +7,31 @@ import { AbstractModel, UnresolvedModel } from "./AbstractModel";
 
 /**
  * Creates an association between the ownerId and its user model
+ * TODO Modify to handle ghost users
  */
 export function Owner<M extends AbstractModel & { ownerId?: Id }>() {
   return HasOne<M>(ACCOUNT, "ownerId");
 }
 
 /**
  * Creates an association between the creatorId and its user model
+ * TODO Modify to handle ghost users
  */
 export function Creator<M extends AbstractModel & { creatorId?: Id }>() {
   return HasOne<M>(ACCOUNT, "creatorId");
 }
 
 /**
  * Creates an association between the updaterId and its user model
+ * TODO Modify to handle ghost users
  */
 export function Updater<M extends AbstractModel & { updaterId?: Id }>() {
   return HasOne<M>(ACCOUNT, "updaterId");
 }
 
 /**
  * Creates an association between the deleterId and its user model
+ * TODO Modify to handle ghost users
```
anonymous users?

Allcharles replied:
deleted/404 ones yes

atruskie commented on src/app/models/Site.ts:
```diff
@@ -107,4 +120,34 @@ export class Site extends AbstractModel implements ISite {
       siteId: this.id,
     });
   }
+
+  /**
+   * Get site latitude
+   */
+  public getLatitude(): number {
+    return this.latitude ?? this.customLatitude;
+  }
+
+  /**
+   * Get site longitude
+   */
+  public getLongitude(): number {
+    return this.longitude ?? this.customLongitude;
+  }
+
```
seems like duplicated logic with the constructor?

Allcharles replied:
Yes and no, the latitude and longitude for a model are only set if the user can actually read the un-obfuscated coordinates. For all other users, they will still use the custom values.

Allcharles replied:
The reason the constructor sets the latitude and longitude to be the same as the custom values is because the API will read the lat/lng values when changing the coordinates of the site. Setting these values for sites where the user does not actually know the correct lat/lng removes the semantic difference provided by the API.

atruskie reviewed (approved):
Looks great.

One note: for the standard image size arrays returned by the API there is really no reason we'd want to show all of the sizes (as per your demo image). I assume you were just testing support for different size images?

Allcharles committed:
Fix unit tests (#362)

* Cleaned up broken unit tests

Had an edge case where it appeared that I could not mock the MapComponent. This turns out to stem from the fact that unit tests cannot override the components for modules which are not directly tied to the component. In this case, the custom formly inputs are loaded as part of a smaller module, which means I cannot override the MapComponent

* Mock services (#363)

* Extract service/guard to modules

* Removed testAppInitializer

* Removed testBawServices

* Mocked services with SpyObjects

* Created ApiErrorDetails generator
